File: backend/scripts/check_licenses.py
from __future__ import annotations

import json
import re
import subprocess
import sys
from dataclasses import dataclass
from datetime import date
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[2]
POLICY_PATH = ROOT / "license-policy.yml"


@dataclass
class ExceptionRule:
    ecosystem: str | None = None
    package: str | None = None
    version: str | None = None
    license: str | None = None
    justification: str | None = None
    owner: str | None = None
    expires_on: date | None = None


@dataclass
class Policy:
    action_on_unknown: str
    fail_on_missing_license: bool
    allowed_licenses: set[str]
    conditionally_allowed_licenses: set[str]
    denied_licenses: set[str]
    exceptions: list[ExceptionRule]


def strip_comment(line: str) -> str:
    in_single = False
    in_double = False
    result: list[str] = []

    for char in line:
        if char == '"' and not in_single:
            in_double = not in_double
        elif char == "'" and not in_double:
            in_single = not in_single
        elif char == "#" and not in_single and not in_double:
            break
        result.append(char)

    return "".join(result).rstrip()


def parse_scalar(value: str) -> Any:
    value = value.strip()
    if not value:
        return ""
    if value == "[]":
        return []
    if value == "{}":
        return {}
    if value.startswith(('"', "'")) and value.endswith(('"', "'")):
        return value[1:-1]
    if value in {"true", "false"}:
        return value == "true"
    if re.fullmatch(r"\d{4}-\d{2}-\d{2}", value):
        return date.fromisoformat(value)
    return value


def parse_policy_file(text: str) -> dict[str, Any]:
    lines = [strip_comment(line) for line in text.splitlines()]
    lines = [line for line in lines if line.strip()]

    root: dict[str, Any] = {}
    stack: list[tuple[int, dict[str, Any] | list[Any]]] = [(-1, root)]

    for index, raw_line in enumerate(lines):
        indent = len(raw_line) - len(raw_line.lstrip(" "))
        line = raw_line.strip()

        while len(stack) > 1 and indent <= stack[-1][0]:
            stack.pop()

        parent = stack[-1][1]

        if line.startswith("- "):
            if not isinstance(parent, list):
                raise ValueError(f"Invalid list item placement: {raw_line}")
            item_text = line[2:].strip()
            if ":" in item_text:
                key, value = item_text.split(":", 1)
                item: dict[str, Any] = {key.strip(): parse_scalar(value)}
                parent.append(item)
                stack.append((indent, item))
            else:
                parent.append(parse_scalar(item_text))
            continue

        key, _, value = line.partition(":")
        key = key.strip()
        value = value.strip()

        if isinstance(parent, list):
            raise ValueError(f"Unexpected mapping under list: {raw_line}")

        if not value:
            next_non_empty = None
            for next_line in lines[index + 1 :]:
                next_indent = len(next_line) - len(next_line.lstrip(" "))
                if next_indent <= indent:
                    break
                next_non_empty = next_line.strip()
                break
            container: dict[str, Any] | list[Any] = (
                [] if next_non_empty and next_non_empty.startswith("- ") else {}
            )
            parent[key] = container
            stack.append((indent, container))
            continue

        parent[key] = parse_scalar(value)

    return root


def load_policy() -> Policy:
    data = parse_policy_file(POLICY_PATH.read_text(encoding="utf-8"))
    defaults = data.get("defaults", {})

    exceptions = []
    for item in data.get("exceptions", []):
        exceptions.append(
            ExceptionRule(
                ecosystem=item.get("ecosystem"),
                package=item.get("package"),
                version=item.get("version"),
                license=item.get("license"),
                justification=item.get("justification"),
                owner=item.get("owner"),
                expires_on=item.get("expires_on"),
            )
        )

    return Policy(
        action_on_unknown=defaults.get("action_on_unknown", "deny"),
        fail_on_missing_license=bool(defaults.get("fail_on_missing_license", True)),
        allowed_licenses=set(data.get("allowed_licenses", [])),
        conditionally_allowed_licenses=set(
            data.get("conditionally_allowed_licenses", [])
        ),
        denied_licenses=set(data.get("denied_licenses", [])),
        exceptions=exceptions,
    )


def normalize_token(token: str) -> str:
    token = token.strip()
    token = re.sub(r"^\((.*)\)$", r"\1", token)
    token = re.sub(r"\s+", " ", token)
    token = token.split("\n", 1)[0].strip()
    if (
        " AND " not in token
        and " OR " not in token
        and ";" not in token
        and "," not in token
    ):
        token = re.sub(r"\s*\([^)]*\)$", "", token).strip()
    mapping = {
        "Apache 2.0": "Apache-2.0",
        "Apache*": "Apache-2.0",
        "Apache Software License": "Apache-2.0",
        "Apache License 2.0": "Apache-2.0",
        "MIT License": "MIT",
        "MIT*": "MIT",
        "BSD License": "BSD-3-Clause",
        "BSD": "BSD-3-Clause",
        "3-Clause BSD License": "BSD-3-Clause",
        "BSD-3-Clause License": "BSD-3-Clause",
        "BSD-2-Clause License": "BSD-2-Clause",
        "The Unlicense": "Unlicense",
        "ISC License": "ISC",
        "Python Software Foundation License": "PSF-2.0",
        "Mozilla Public License 2.0": "MPL-2.0",
        "PSF": "PSF-2.0",
        "ZLIB": "Zlib",
        "UNLICENSED": "UNLICENSED",
    }
    return mapping.get(token, token)


def split_expression(license_value: str) -> tuple[str, list[str]]:
    cleaned = re.sub(r"\s+", " ", license_value.strip())
    if not cleaned:
        return "single", []

    inner = cleaned.strip("()").replace(";", " OR ").replace(",", " OR ")
    if " OR " in inner:
        return "or", [
            normalize_token(part) for part in inner.split(" OR ") if part.strip()
        ]
    if " AND " in inner:
        return "and", [
            normalize_token(part) for part in inner.split(" AND ") if part.strip()
        ]
    if "/" in inner and "http" not in inner:
        return "or", [
            normalize_token(part) for part in inner.split("/") if part.strip()
        ]
    return "single", [normalize_token(inner)]


def is_exception_match(
    rule: ExceptionRule, package: str, version: str, license_value: str
) -> bool:
    if rule.ecosystem and rule.ecosystem != "python":
        return False
    if rule.package and rule.package != package:
        return False
    if rule.version and rule.version != version:
        return False
    if rule.license and rule.license != normalize_token(license_value):
        return False
    if rule.expires_on and rule.expires_on < date.today():
        return False
    return True


def find_exception(
    policy: Policy, package: str, version: str, license_value: str
) -> ExceptionRule | None:
    normalized = normalize_token(license_value)
    for rule in policy.exceptions:
        if is_exception_match(rule, package, version, normalized):
            return rule
    return None


def evaluate_license(policy: Policy, license_value: str) -> tuple[bool, str]:
    normalized = normalize_token(license_value)
    if not normalized or normalized.upper() == "UNKNOWN":
        if policy.fail_on_missing_license or policy.action_on_unknown == "deny":
            return False, "unknown license"
        return True, "unknown license allowed by policy"

    mode, parts = split_expression(normalized)
    if not parts:
        return False, "missing license"

    def classify(token: str) -> str:
        if token in policy.denied_licenses:
            return "denied"
        if (
            token in policy.allowed_licenses
            or token in policy.conditionally_allowed_licenses
        ):
            return "allowed"
        return "unknown"

    statuses = [classify(part) for part in parts]

    if mode == "or":
        if "allowed" in statuses:
            return True, f"allowed via OR expression: {', '.join(parts)}"
        if "denied" in statuses:
            return False, f"denied OR expression: {', '.join(parts)}"
        return False, f"unknown OR expression: {', '.join(parts)}"

    if mode == "and":
        if all(status == "allowed" for status in statuses):
            return True, f"allowed AND expression: {', '.join(parts)}"
        denied = [part for part, status in zip(parts, statuses) if status == "denied"]
        if denied:
            return False, f"denied AND expression component: {', '.join(denied)}"
        unknown = [part for part, status in zip(parts, statuses) if status == "unknown"]
        return False, f"unknown AND expression component: {', '.join(unknown)}"

    token = parts[0]
    if token in policy.denied_licenses:
        return False, f"denied license: {token}"
    if token in policy.allowed_licenses:
        return True, f"allowed license: {token}"
    if token in policy.conditionally_allowed_licenses:
        return True, f"conditionally allowed license: {token}"
    return False, f"unknown license: {token}"


def run_pip_licenses() -> list[dict[str, Any]]:
    result = subprocess.run(
        ["pip-licenses", "--format=json", "--with-urls"],
        check=True,
        capture_output=True,
        text=True,
    )
    return json.loads(result.stdout)


def main() -> int:
    policy = load_policy()
    packages = run_pip_licenses()
    failures: list[str] = []

    for package in packages:
        name = package.get("Name", "")
        version = package.get("Version", "")
        license_value = str(package.get("License", "")).strip()
        license_value = license_value.split("\n", 1)[0].strip()

        exception = find_exception(policy, name, version, license_value)
        if exception is not None:
            continue

        allowed, reason = evaluate_license(policy, license_value)
        if not allowed:
            failures.append(
                f"- {name}=={version}: {license_value or 'UNKNOWN'} ({reason})"
            )

    if failures:
        print("License compliance check failed for backend dependencies:")
        print("\n".join(failures))
        return 1

    print(f"Backend license compliance check passed for {len(packages)} dependencies.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: frontend/scripts/check-licenses.mjs
import { readFile } from 'node:fs/promises'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { execFile } from 'node:child_process'
import { promisify } from 'node:util'

const execFileAsync = promisify(execFile)
const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const rootDir = path.resolve(__dirname, '..', '..')
const frontendDir = path.resolve(__dirname, '..')
const policyPath = path.join(rootDir, 'license-policy.yml')
const nextConfigPath = path.join(frontendDir, 'next.config.ts')

function stripComment(line) {
  let inSingle = false
  let inDouble = false
  let result = ''

  for (const char of line) {
    if (char === '"' && !inSingle) {
      inDouble = !inDouble
    } else if (char === "'" && !inDouble) {
      inSingle = !inSingle
    } else if (char === '#' && !inSingle && !inDouble) {
      break
    }
    result += char
  }

  return result.trimEnd()
}

function parseScalar(value) {
  const trimmed = value.trim()
  if (!trimmed) return ''
  if (trimmed === '[]') return []
  if (trimmed === '{}') return {}
  if ((trimmed.startsWith('"') && trimmed.endsWith('"')) || (trimmed.startsWith("'") && trimmed.endsWith("'"))) {
    return trimmed.slice(1, -1)
  }
  if (trimmed === 'true') return true
  if (trimmed === 'false') return false
  return trimmed
}

function parsePolicy(text) {
  const lines = text.split('\n').map(stripComment).filter((line) => line.trim())
  const root = {}
  const stack = [{ indent: -1, value: root }]

  for (let index = 0; index < lines.length; index += 1) {
    const rawLine = lines[index]
    const indent = rawLine.length - rawLine.trimStart().length
    const line = rawLine.trim()

    while (stack.length > 1 && indent <= stack[stack.length - 1].indent) {
      stack.pop()
    }

    const parent = stack[stack.length - 1].value

    if (line.startsWith('- ')) {
      if (!Array.isArray(parent)) {
        throw new Error(`Invalid list item placement: ${rawLine}`)
      }

      const itemText = line.slice(2).trim()
      if (itemText.includes(':')) {
        const separatorIndex = itemText.indexOf(':')
        const key = itemText.slice(0, separatorIndex).trim()
        const value = itemText.slice(separatorIndex + 1).trim()
        const item = { [key]: parseScalar(value) }
        parent.push(item)
        stack.push({ indent, value: item })
      } else {
        parent.push(parseScalar(itemText))
      }
      continue
    }

    const separatorIndex = line.indexOf(':')
    const key = line.slice(0, separatorIndex).trim()
    const value = line.slice(separatorIndex + 1).trim()

    if (Array.isArray(parent)) {
      throw new Error(`Unexpected mapping under list: ${rawLine}`)
    }

    if (!value) {
      let nextNonEmpty = null
      for (let nextIndex = index + 1; nextIndex < lines.length; nextIndex += 1) {
        const nextLine = lines[nextIndex]
        const nextIndent = nextLine.length - nextLine.trimStart().length
        if (nextIndent <= indent) {
          break
        }
        nextNonEmpty = nextLine.trim()
        break
      }
      const container = nextNonEmpty && nextNonEmpty.startsWith('- ') ? [] : {}
      parent[key] = container
      stack.push({ indent, value: container })
      continue
    }

    parent[key] = parseScalar(value)
  }

  return root
}

function normalizeToken(token) {
  const cleaned = token
    .trim()
    .replace(/^\(+|\)+$/g, '')
    .replace(/\s+/g, ' ')
    .split('\n', 1)[0]
    .replace(/\s*\([^)]*\)$/g, '')
    .trim()
  const mapping = {
    'Apache 2.0': 'Apache-2.0',
    'Apache*': 'Apache-2.0',
    'Apache License 2.0': 'Apache-2.0',
    'Apache Software License': 'Apache-2.0',
    'MIT License': 'MIT',
    'MIT*': 'MIT',
    'BSD': 'BSD-3-Clause',
    'BSD License': 'BSD-3-Clause',
    '3-Clause BSD License': 'BSD-3-Clause',
    'BSD-3-Clause License': 'BSD-3-Clause',
    'BSD-2-Clause License': 'BSD-2-Clause',
    'The Unlicense': 'Unlicense',
    'ISC License': 'ISC',
    'Python Software Foundation License': 'PSF-2.0',
    'Mozilla Public License 2.0': 'MPL-2.0',
    'PSF': 'PSF-2.0',
    'ZLIB': 'Zlib',
    'UNLICENSED': 'UNLICENSED',
  }
  return mapping[cleaned] ?? cleaned
}

function splitExpression(licenseValue) {
  const cleaned = licenseValue.trim().replace(/\s+/g, ' ')
  if (!cleaned) return { mode: 'single', parts: [] }

  const inner = cleaned.replace(/^\(+|\)+$/g, '').replaceAll(';', ' OR ').replaceAll(',', ' OR ')
  if (inner.includes(' OR ')) {
    return { mode: 'or', parts: inner.split(' OR ').map(normalizeToken).filter(Boolean) }
  }
  if (inner.includes(' AND ')) {
    return { mode: 'and', parts: inner.split(' AND ').map(normalizeToken).filter(Boolean) }
  }
  if (inner.includes('/') && !inner.includes('http')) {
    return { mode: 'or', parts: inner.split('/').map(normalizeToken).filter(Boolean) }
  }
  return { mode: 'single', parts: [normalizeToken(inner)] }
}

function classify(policy, token) {
  if (policy.denied.has(token)) return 'denied'
  if (policy.allowed.has(token) || policy.conditional.has(token)) return 'allowed'
  return 'unknown'
}

function evaluateLicense(policy, licenseValue) {
  const normalized = normalizeToken(licenseValue)
  if (!normalized || normalized.toUpperCase() === 'UNKNOWN') {
    if (policy.failOnMissingLicense || policy.actionOnUnknown === 'deny') {
      return { ok: false, reason: 'unknown license' }
    }
    return { ok: true, reason: 'unknown license allowed by policy' }
  }

  const { mode, parts } = splitExpression(normalized)
  if (parts.length === 0) {
    return { ok: false, reason: 'missing license' }
  }

  const statuses = parts.map((part) => classify(policy, part))

  if (mode === 'or') {
    if (statuses.includes('allowed')) {
      return { ok: true, reason: `allowed via OR expression: ${parts.join(', ')}` }
    }
    if (statuses.includes('denied')) {
      return { ok: false, reason: `denied OR expression: ${parts.join(', ')}` }
    }
    return { ok: false, reason: `unknown OR expression: ${parts.join(', ')}` }
  }

  if (mode === 'and') {
    if (statuses.every((status) => status === 'allowed')) {
      return { ok: true, reason: `allowed AND expression: ${parts.join(', ')}` }
    }
    const denied = parts.filter((part, index) => statuses[index] === 'denied')
    if (denied.length > 0) {
      return { ok: false, reason: `denied AND expression component: ${denied.join(', ')}` }
    }
    const unknown = parts.filter((part, index) => statuses[index] === 'unknown')
    return { ok: false, reason: `unknown AND expression component: ${unknown.join(', ')}` }
  }

  const token = parts[0]
  if (policy.denied.has(token)) return { ok: false, reason: `denied license: ${token}` }
  if (policy.allowed.has(token)) return { ok: true, reason: `allowed license: ${token}` }
  if (policy.conditional.has(token)) return { ok: true, reason: `conditionally allowed license: ${token}` }
  return { ok: false, reason: `unknown license: ${token}` }
}

function isExceptionExpired(expiresOn) {
  if (!expiresOn) return false
  return new Date(`${expiresOn}T00:00:00Z`) < new Date(new Date().toISOString().slice(0, 10) + 'T00:00:00Z')
}

function findException(policy, packageName, version, licenseValue) {
  const normalized = normalizeToken(licenseValue)
  return policy.exceptions.find((rule) => {
    if (rule.ecosystem && rule.ecosystem !== 'node') return false
    if (rule.package && rule.package !== packageName) return false
    if (rule.version && rule.version !== version) return false
    if (rule.license && rule.license !== normalized) return false
    if (isExceptionExpired(rule.expires_on)) return false
    return true
  })
}

async function loadPolicy() {
  const text = await readFile(policyPath, 'utf-8')
  const data = parsePolicy(text)
  return {
    actionOnUnknown: data.defaults?.action_on_unknown ?? 'deny',
    failOnMissingLicense: Boolean(data.defaults?.fail_on_missing_license ?? true),
    allowed: new Set(data.allowed_licenses ?? []),
    conditional: new Set(data.conditionally_allowed_licenses ?? []),
    denied: new Set(data.denied_licenses ?? []),
    exceptions: data.exceptions ?? [],
  }
}

async function runLicenseChecker() {
  const { stdout } = await execFileAsync(
    path.join(frontendDir, 'node_modules', '.bin', 'license-checker-rseidelsohn'),
    ['--json', '--production'],
    { cwd: frontendDir, maxBuffer: 10 * 1024 * 1024 },
  )
  return JSON.parse(stdout)
}

async function isNextImageOptimizationDisabled() {
  const config = await readFile(nextConfigPath, 'utf-8')
  return /images\s*:\s*\{[\s\S]*?unoptimized\s*:\s*true/.test(config)
}

function shouldIgnorePackage(name, disableNextImageOptimization) {
  if (!disableNextImageOptimization) {
    return false
  }

  return name === 'sharp' || name.startsWith('@img/sharp-') || name.startsWith('@img/sharp-libvips-')
}

function parsePackageKey(key) {
  const separatorIndex = key.lastIndexOf('@')
  return {
    name: key.slice(0, separatorIndex),
    version: key.slice(separatorIndex + 1),
  }
}

async function main() {
  const policy = await loadPolicy()
  const packages = await runLicenseChecker()
  const disableNextImageOptimization = await isNextImageOptimizationDisabled()
  const failures = []
  let ignoredPackages = 0

  for (const [key, info] of Object.entries(packages)) {
    const { name, version } = parsePackageKey(key)
    if (name === 'clouisle') {
      continue
    }
    if (shouldIgnorePackage(name, disableNextImageOptimization)) {
      ignoredPackages += 1
      continue
    }

    const licenseValue = String(info.licenses ?? '').trim().split('\n', 1)[0].trim()

    if (findException(policy, name, version, licenseValue)) {
      continue
    }

    const result = evaluateLicense(policy, licenseValue)
    if (!result.ok) {
      failures.push(`- ${name}@${version}: ${licenseValue || 'UNKNOWN'} (${result.reason})`)
    }
  }

  if (failures.length > 0) {
    console.error('License compliance check failed for frontend dependencies:')
    console.error(failures.join('\n'))
    process.exit(1)
  }

  const checkedCount = Object.keys(packages).length - ignoredPackages
  if (ignoredPackages > 0) {
    console.log(`Frontend license compliance check passed for ${checkedCount} dependencies. Ignored ${ignoredPackages} optional Next image optimization packages because images.unoptimized is enabled.`)
    return
  }

  console.log(`Frontend license compliance check passed for ${checkedCount} dependencies.`)
}

main().catch((error) => {
  console.error(error instanceof Error ? error.message : String(error))
  process.exit(1)
})
